src/Bitmap_cubical_complex/include/gudhi/Persistence_on_rectangle.h

```diff
@@ -0,0 +1,595 @@
+    while (parent != grandparent)
+    {
+      ds_parent(v) = grandparent;
+      v = parent;
+      parent = grandparent;
+      grandparent = ds_parent(parent);
+    }
+    return parent;
+#elif 1
+    // No compression (just for reference)
+    Index parent;
+    while (v != (parent = ds_parent(v)))
+      v = parent;
+    return v;
+#endif
+  }
+  Index ds_find_set_vertex(Index v) {
+    return ds_find_set_(v, [this](Index i) -> Index& { return ds_parent_vertex(i); });
+  }
+  Index ds_find_set_square(Index v) {
+    return ds_find_set_(v, [this](Index i) -> Index& { return ds_parent_square(i); });
+  }
+
+  struct Edge {
+    T f;
+    Index v1, v2; // v1 < v2
+    Edge() = default;
+    Edge(T f, Index v1, Index v2) : f(f), v1(v1), v2(v2) {}
+    Filtration_value filt() const { return f.first; }
+    bool operator<(Edge const& other) const { return filt() < other.filt(); }
+  };
+  void dualize_edge(Edge& e) const {
+    Index new_v2 = e.v1 + (dy + 1);
+    e.v1 = e.v2;
+    e.v2 = new_v2;
+  };
+  std::vector<Edge> edges;
+
+  void init(const Filtration_value* input_, Index n_rows, Index n_cols) {
+    input_size = n_rows * n_cols;
+    input_p = input_;
+#ifdef DEBUG_TRACES
+    std::clog << "Input\n";
+    for(Index i = 0; i < input_size; ++i) {
+      std::clog << i << '\t' << input(i) << '\n';
+    }
+#endif
+    dy = n_cols;
+    size_x = dy - 1;
+    size_y = n_rows - 1;
+    // The unique_ptr could be std::vector, but the initialization is useless.
+    data_v_.reset(new T[input_size - dy - 1]); // 1 row/column less for vertices than squares
+    ds_parent_v_.reset(new Index[input_size - dy - 1]);
+    ds_parent_s_.resize(input_size); // Initializing the boundary squares to 0 is important, it represents the infinite exterior cell.
+    edges.reserve(input_size / 2); // What is a good estimate here? For a random 1000x1000 input, we get ~311k edges. For a checkerboard, ~498k.
+  }
+
+  bool has_larger_input(Index a, Index b, Filtration_value fb) const {
+    // Is passing fb useful, or would the compiler notice that it already has it available?
+    GUDHI_CHECK(a != b, std::logic_error("Bug in Gudhi"));
+    Filtration_value fa = input(a);
+    if (fb < fa) return true;
+    if (fa < fb) return false;
+    return a > b; // Arbitrary, but has to be consistent
+  }
+  void set_parent_vertex(Index child, Index parent) {
+    GUDHI_CHECK(child != parent, std::logic_error("Bug in Gudhi: use mark_*_critical instead of set_parent"));
+    ds_parent_vertex(child) = parent;
+  }
+  void set_parent_square(Index child, Index parent) {
+    GUDHI_CHECK(child != parent, std::logic_error("Bug in Gudhi: use mark_*_critical instead of set_parent"));
+    ds_parent_square(child) = parent;
+  }
+
+  // Locally pair simplices around each square.
+  // Work implicitly from input, only store the filtration value of critical vertices (squares are already in input).
+  // Store critical edges for later processing.
+  void fill_and_pair() {
```

Comment:
Is it possible to decompose this method to have a better overview?

Reply:
I am not sure how to do that in a useful way.
The function starts by setting up the vocabulary: `i` is the current square, its upper-left vertex is referred to as `v_ul`, etc. Then it iterates on all the squares. And the scary part, well-delimited in its for loop, is the case analysis for interior squares, which depends on which neighboring squares have a bigger / smaller filtration value.

I could split out the handling of a single interior square into its own function, although I would then have to duplicate the vocabulary lambdas, but I am not sure it would be much clearer.

Do you have some precise suggestion?

Reply:
By "overview", do you mean you would like the content of the for loops to be short enough that we can see them completely on one screen?
A comment
```
// Iterate on all the squares
// For each square in the first row: ...
// For each row:
//     First square
//     For each interior square: ...
//     Last square
// For each square in the last row: ...
```
could play the same role?

Reply:
I see what you mean. It is just that there are so many imbricated `if`s and small letters everywhere, that I always get lost at some point. I am wondering if replacing `u`, `d`, `r`, `l` by their respective meanings would not make them easier to read (even though I agree that the meanings are clear).
Thankfully, the collapsing function of the editor is helping a lot already!

Reply:
No, the little functions are a very good idea! 
I meant `r` ---> `right`, `l`---> `left` etc.
If I wrote all those imbricated conditions, I wouldn't want to touch them again either. And if it is the fastest way, they should be left like that anyway.

Reply:
I just pushed a patch that renames to longer names. What do you think?
(I am not convinced that it is more readable, but I am ok with it if you prefer this)

Reply:
Sorry, I answer two month later, but yes, I prefer the new version.